spec/fiber/local/inheritable_spec.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+# Copyright, 2021, by Samuel G. D. Williams. <http://www.codeotaku.com>
+# 
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+# 
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+# 
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+require 'fiber/local/inheritable'
+
+class MyFiber < Fiber
+	prepend Fiber::Local::Inheritable
+	
+	attr_accessor :id
+	inheritable :id
```

Comment:
I still feel this is very confusing, whether a "fiber local" is inherited should depend on what creates the Fiber for what purpose (i.e., is it a "nested Fiber"), and not on the attribute.

Why not something like `inheritable_fiber_local :id` or so? (without the `attr_accessor :id`)

Reply:
The entire point of this is to allow fiber locals to inherit across uncontrolled boundaries, like say Enumerator's `Fiber.new` or the schedulers `Fiber.new`. It should be inherited by default for things like `request_id` so that the same request_id is associated with any child work/fibers.

To make at lest some way to opt out, we might consider adding some option like `Fiber.new(inherit: false)`.

Regarding separating out `accessor` from `inheritable` is because they are two different things and you might have custom accessors but you still want the instance variable to be inherited.

Reply:
> The entire point of this is to allow fiber locals to inherit across uncontrolled boundaries, like say Enumerator's `Fiber.new` or the schedulers `Fiber.new`. It should be inherited by default for things like `request_id` so that the same request_id is associated with any child work/fibers.

I see. Changing the Fiber.new in the scheduler seems something feasible, but for Enumerator it would need CRuby changes and a new build/release or so, which is less convenient. OTOH I guess people rarely if ever use Enumerator fibers (it's slow and usually not needed).

The other side of this is this new API would work for whoever uses it, but won't work for all existing usages of Fiber locals, which sounds a very serious limitation. That makes me think copying Fiber locals in the scheduler is much more useful in practice, but maybe I'm wrong?

Reply:
I think your assessment is largely correct.

> The other side of this is this new API would work for whoever uses it, but won't work for all existing usages of Fiber locals, which sounds a very serious limitation

I actually see this as an advantage because it's a new semantic and we don't want to break existing usage of fiber locals.

>  That makes me think copying Fiber locals in the scheduler and (and other places creating "nested Fibers") is much more useful in practice, but maybe I'm wrong?

I think that's a fair assessment but I'm not sure why we need to be so specific when a general approach would work i.e. in `Fiber.new` everywhere by default for instance variables that are declared to be inherited.

Reply:
> OTOH I guess people rarely if ever use Enumerator fibers (it's slow and usually not needed).

`to_enum` usually creates a fiber, right?

Reply:
> `to_enum` usually creates a fiber, right?

No, it creates an Enumerator but a Fiber is only created and needed for `Enumerator#{next*,peek*}`, which I would think is extremely rarely used.

Reply:
Ah, got it. Yes, I sometimes use `#next` but you are right it's probably uncommon.